### PR TITLE
SPI Optimisations

### PR DESCRIPTION
**Pull request by pkendall64**
Optimisations of the SPI interface for ESP-based devices.

1. Use hardware controlled CS/NSS pin on ESP8285 devices (big gains)
2. Upgrade SPI clock to 17.5MHz (small gains but noticeable)
3. Avoid some memory copies (small gains)
4. Extend the SPI implementation class to provide out own low-level transfer function
   a. After performing a write operation do not wait for it to complete, the processor has better things to do (decent gains)
   b. Wait for SPI to be non-busy when starting an SPI operation
   c. Don't do individual byte operations when the size is not word-boundary sized (i.e. modulo 4) as the buffer is padded to word-boundary size, so it can be optimised. (good gains)
5. Move CS control into the transfer function
   a. with notes that dual radio is only supported in ESP32 devices

# Logic analyser traces

## ESP8285 on current master
8285-master 2

## ESP8285 on this branch
8285-nowait 2

As can be seen from the above traces, hardware CS control removes the 20us CS initialisation, not sure why it's so big at the start but there you go.
Also, the inter-operation time is shrunk because the processor can continue execution while the SPI data is being transferred.
You can also see the effect of the non-word boundary operations where the (final) 3 bytes are transferred independantly on the old driver.

## ESP32 (dual) on current master
esp32-master

## ESP32 (dual) on this branch
esp32-nowait

### CODE REVIEW

**Review by JyeSmith — APPROVED** [review]
Love it 💪 

Tested that it is working operationally on an 8285 Emax rx and ESP32 Gemini rx.  Lets get this merged and Ill have a closer look at the SPI when working on the LR1121.

**Comment by JyeSmith** [comment]
Looking good and Ive been testing it over the past few days.  But man is it a PITA to log all the channels I need at 17.5M 😭

**Comment by JyeSmith** [comment]
Minor change in CS to prevent their state from being changed mid packet.

image

**Comment by pkendall64** [comment]
Yup, than last change makes good sense.

File: src/lib/SPIEx/SPIEx.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#include "targets.h"
+#include <SPI.h>
+
+/**
+ * @brief An extension to the platform SPI class then provides som performance enhancements.
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:5]: that
some

File: src/lib/SPIEx/SPIEx.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * 1. The provided data buffer is expected to be word-aligned and word-padded, this allow the
+ *    data to be copied to the SPI fifo un word-sized chunks.
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:8]: in

File: src/lib/SPIEx/SPIEx.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#if defined(PLATFORM_ESP32)
+    explicit SPIExClass(uint8_t spi_bus=HSPI) : SPIClass(spi_bus) {}
+#else
+    explicit SPIExClass() : SPIClass() {}
+#endif
+
+    /**
+     * @brief Perform an SPI read operation on the SPI bus.
+     *
+     * If the SPI bus is busy the processor wait for the current operation to complete before starting
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:25]: waits

File: src/lib/SPIEx/SPIEx.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+     * @param cs_mask mask of CS pins to enable for this operation
+     * @param data word-aligned and padded data buffer
+     * @param size the number of bytes to be read into in the data buffer
+     */
+    void inline ICACHE_RAM_ATTR read(uint8_t cs_mask, uint8_t *data, uint32_t size) { _transfer(cs_mask, data, size, true); }
+
+    /**
+     * @brief Perform an SPI write operation on the SPI bus.
+     *
+     * If the SPI bus is busy the processor wait for the current operation to complete before starting
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:39]: waits

File: src/lib/SPIEx/SPIEx.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+     * this operation.
+     * One the SPI bus is not busy, it copies the data to the SPI fifo then returns without waiting for
+     * the operation to complete allowing the processor to do other work.
+     *
+     * If the SPI bus is busy at the when start this operation starts,
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:44]: at the when?

File: src/lib/SPIEx/SPIEx.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+     *
+     * @param cs_mask mask of CS pins to enable for this operation
+     * @param data word-aligned and padded data buffer
+     * @param size the number of bytes to be write to the SPI device
```
Comment (wvarty) [inline src/lib/SPIEx/SPIEx.h:48]: to be written?

**Review by wvarty — APPROVED** [review]

**Pull request merged**